plugins/rbac/src/components/CreateRole/RoleForm.tsx

```diff
@@ -17,35 +17,71 @@ import { Alert } from '@material-ui/lab';
 import { FormikHelpers, useFormik } from 'formik';
 
 import { rbacApiRef } from '../../api/RBACBackendClient';
+import { MemberEntity } from '../../types';
 import { getRoleData, validationSchema } from '../../utils/create-role-utils';
+import { useToast } from '../ToastContext';
 import { AddedMembersTable } from './AddedMembersTable';
 import { AddMembersForm } from './AddMembersForm';
 import { RoleDetailsForm } from './RoleDetailsForm';
-import { CreateRoleFormValues } from './types';
+import { RoleFormValues } from './types';
 
-export const CreateRoleForm = () => {
-  const [activeStep, setActiveStep] = React.useState<number>(0);
+type RoleFormProps = {
+  membersData: { members: MemberEntity[]; loading: boolean; error: Error };
+  titles: {
+    formTitle: string;
+    nameAndDescriptionTitle: string;
+    usersAndGroupsTitle: string;
+    permissionPoliciesTitle: string;
+  };
+  submitLabel?: string;
+  roleName?: string;
+  step?: number;
+  initialValues: RoleFormValues;
+};
+
+export const RoleForm = ({
+  roleName,
+  step,
+  titles,
+  membersData,
+  submitLabel,
+  initialValues,
+}: RoleFormProps) => {
+  const { setToastMessage } = useToast();
+  const [activeStep, setActiveStep] = React.useState<number>(step || 0);
   const navigate = useNavigate();
   const rbacApi = useApi(rbacApiRef);
-  const formik = useFormik<CreateRoleFormValues>({
-    initialValues: {
-      name: '',
-      namespace: 'default',
-      description: '',
-      selectedMembers: [],
-    },
+
+  const formik = useFormik<RoleFormValues>({
+    enableReinitialize: true,
+    initialValues,
     validationSchema: validationSchema,
     onSubmit: async (
-      values: CreateRoleFormValues,
-      formikHelpers: FormikHelpers<CreateRoleFormValues>,
+      values: RoleFormValues,
+      formikHelpers: FormikHelpers<RoleFormValues>,
     ) => {
       try {
-        const data = getRoleData(values);
-        const res = await rbacApi.createRole(data);
+        const newData = getRoleData(values);
+        const oldData = getRoleData(initialValues);
+
+        let res: Response;
```

Comment:
I wonder if this block of code dealing with the `Response` directly should be moved to the API layer.  Usually it's best to avoid having to deal with this at the UI component level.  

As an example I'm a big fan of the approach that was taken [here](https://github.com/bf2fc6cc711aee1a0c2a/cos-ui/blob/main/src/apis/api.ts#L316-L366) when this was first developed, success/error callbacks are provided to the API function and those API call details are kept away from the UI component.  The caller instead provides the token and details to access the server and gets back a function that provides onSuccess and onError callbacks, one of these are called depending on the result of the remote call.  Not saying we adopt this approach now or at all, just an example of what I've seen work well.

Reply:
Thank you for pointing this out, I handled this in the other API functions but overlooked it here. I couldn't create scenarios where the API call would throw an error. Even with an incorrect bearer token or query parameters, it would still enter the "then" block. I could only display the error to the user by reading the status code in the response. PTAL at the updated API functions.

Reply:
Ah yes, with the `fetch` API it doesn't throw a TypeError for some HTTP error return codes.  You instead have to handle it directly like:

```
fetch(someUrl).then(( response:Response ) => {
  if ( !response.ok ) {
    throw new Error(response.message); // or maybe a more informative string based on Response.status
  }
  // happy path here
}).catch(( err:Error ) => {
  console.log("Something bad happened: ", err);
});
```

That snippet I sent along used axios under the covers which does this already and hides the above.

Reply:
Another option with this section is maybe try and first just move the logic to a separate function that accepts `onSuccess` & `onError` callback functions.  Instead of returning a value or throwing an exception the function would instead either call `onSuccess` and supply the `roleName` as an argument or `onError` and supply the remote error as an argument.  This works nicely especially when there's multiple conditions to deal with like you have here.